dali/pipeline/operator/arg_helper.h

```diff
@@ -15,96 +15,95 @@
+
+  ArgValue(const std::string &arg_name, const OpSpec &spec)
+      : arg_name_(arg_name),
+        has_arg_const_(spec.HasArgument(arg_name)),
+        has_arg_input_(spec.HasTensorArgument(arg_name)) {
+    assert(!(has_arg_const_ && has_arg_input_));
   }
 
-  ArgValue &operator=(ArgValue &&) = default;
-  inline ArgValue &operator=(const ArgValue &other) {
-    name_ = other.name_;
-    is_set_ = other.is_set_;
-    gpu_dirty_ = true;
-    gpu_.reset();
-    value_  = other.value_;
-    arg_input_ = other.arg_input_;
-    return *this;
+  bool IsDefined() const {
+    return has_arg_const_ || has_arg_input_;
   }
 
-  inline bool IsInput() const noexcept { return arg_input_ != nullptr; }
-  inline bool IsSet() const noexcept { return is_set_; }
+  bool IsConstant() const {
+    return has_arg_const_;
+  }
 
-  template <int ndim = DynamicDimensions>
-  inline const TensorView<StorageCPU, T, ndim> Tensor(int sample_index) const {
-    ASSERT(IsInput());
-    return view<T, ndim>(arg_input_[sample_index]);
+  bool IsArgInput() const {
+    return has_arg_input_;
   }
 
-  inline const T &Value() const {
-    return value_;
+  void Acquire(const OpSpec &spec, const ArgumentWorkspace &ws, int nsamples = 1,
+               bool enforce_uniform = false) {
+    if (has_arg_input_) {
+      view_ = view<const T, ndim>(ws.ArgumentInput(arg_name_));
+      if (enforce_uniform) {
+        DALI_ENFORCE(is_uniform(view_.shape),
+          make_string("Expected uniform shape for argument \"", arg_name_,
+                      "\" but got shape ", view_.shape));
+      }
+    } else if (has_arg_const_) {
+      TensorShape<ndim> shape{};
+      if (ndim == 1) {
+        data_ = spec.GetRepeatedArgument<T>(arg_name_);
+        int64_t len = data_.size();
+        shape = std::array<int64_t, 1>{len};
+      } else if (ndim == 0) {
+        data_ = {spec.GetArgument<T>(arg_name_)};
+      } else {
+        assert(false);
+      }
+      std::vector<const T*> ptrs(nsamples, nullptr);
+      for (int i = 0; i < nsamples; i++)
+        ptrs[i] = data_.data();
+      view_ = TLV(std::move(ptrs), uniform_list_shape(nsamples, shape));
+    }
   }
 
-  inline const string &Name() const noexcept {
-    return name_;
+  const std::string &name() const {
+    return arg_name_;
   }
 
-  inline const T &operator[](int sample_index) const {
-    if (IsInput()) {
-#if DALI_DEBUG
-      DALI_ENFORCE(sample_index < static_cast<int>(arg_input_->ntensor()));
-#endif
-      return (*arg_input_)[sample_index].data<T>()[0];
-    } else {
-      return  value_;
-    }
+  TV operator[](size_t idx) const {
+    assert(idx < size());
+    return view_[idx];
   }
 
-  inline const TensorList<GPUBackend> &AsGPU(cudaStream_t stream) {
-    DALI_ENFORCE(IsInput());
-    if (!gpu_)
-      gpu_.reset(new TensorList<GPUBackend>());
-    if (gpu_dirty_) {
-      gpu_->Copy(*arg_input_, stream);
-      gpu_dirty_ = false;
-    }
-    return *gpu_;
+  size_t size() const {
```

Comment:
```suggestion
  int size() const {
```
That's what we have in TensorListView and TensorListShape.